core/shared/src/main/scala/zio/stm/TArray.scala

```diff
@@ -39,10 +98,124 @@ class TArray[A] private (val array: Array[TRef[A]]) extends AnyVal {
+    if (from >= 0) forIndex(from) else STM.succeed(-1)
+  }
+
+  /** Get the index of the first entry in the array matching an STM predicate. */
+  final def indexWhereM[E](predicate: A => STM[E, Boolean]): STM[E, Int] = indexWhereM(predicate, 0)
+
+  /** Get the index of the first entry in the array, starting at a specific index, matching an STM predicate. */
+  final def indexWhereM[E](predicate: A => STM[E, Boolean], from: Int): STM[E, Int] = {
+    val len = array.length
+    def forIndex(i: Int): STM[E, Int] =
+      if (i >= len) STM.succeed(-1)
+      else apply(i).flatMap(a => predicate(a).flatMap(result => if (result) STM.succeed(i) else forIndex(i + 1)))
+
+    if (from >= 0) forIndex(from) else STM.succeed(-1)
+  }
+
+  /** The last entry in the array. */
+  final def last: STM[Nothing, A] =
+    if (array.isEmpty) STM.die(new NoSuchElementException("last of empty array"))
+    else array.last.get
+
+  /** Get the last index of a specific value in the array or -1 if it does not occur. */
+  final def lastIndexOf(a: A): STM[Nothing, Int] =
+    if (array.isEmpty) STM.succeed(-1) else lastIndexOf(a, array.length - 1)
+
+  /** Get the first index of a specific value in the array, bounded above by a specific index, or -1 if it does not occur. */
+  final def lastIndexOf(a: A, end: Int): STM[Nothing, Int] = {
+    def forIndex(i: Int): STM[Nothing, Int] =
+      if (i < 0) STM.succeed(-1)
+      else apply(i).flatMap(ai => if (ai == a) STM.succeed(i) else forIndex(i - 1))
+
+    if (end < array.length) forIndex(end) else STM.succeed(-1)
+  }
+
+  /** The last entry in the array, if it exists. */
+  final def lastOption: STM[Nothing, Option[A]] = if (array.isEmpty) STM.succeed(None) else array.last.get.map(Some(_))
+
+  /** Atomically compute he greatest element in the array. */
+  final def max(implicit ord: Ordering[A]): STM[Nothing, A] =
+    reduce((acc, a) => if (ord.gt(a, acc)) a else acc)
+
+  /** Atomically compute the greatest element in the array, if it exists. */
+  final def maxOption(implicit ord: Ordering[A]): STM[Nothing, Option[A]] =
+    reduceOption((acc, a) => if (ord.gt(a, acc)) a else acc)
+
+  /** Atomically compute the least element in the array. */
+  final def min(implicit ord: Ordering[A]): STM[Nothing, A] =
+    reduce((acc, a) => if (ord.lt(a, acc)) a else acc)
+
+  /** Atomically compute the least element in the array, if it exists. */
+  final def minOption(implicit ord: Ordering[A]): STM[Nothing, Option[A]] =
+    reduceOption((acc, a) => if (ord.lt(a, acc)) a else acc)
+
+  /** Atomically reduce the array by a binary operator. */
+  final def reduce(op: (A, A) => A): STM[Nothing, A] =
+    if (array.isEmpty) STM.die(new UnsupportedOperationException("reduction of empty array"))
+    else
+      array.head.get
+        .flatMap(h => new TArray(array.tail).fold(h)((m, a) => op(m, a)))
+
+  /** Atomically reduce the array by an STM binary operator. */
+  final def reduceM[E](op: (A, A) => STM[E, A]): STM[E, A] =
```

Comment:
I would nuke this one and `reduce`, because they can blow up. The safer versions are much nicer to work with!

Reply:
I agree entirely; I was just trying to stay as close to Seq as possible. More than happy to get rid of them!